From 3cec838b066cd5232aa27aaa9da5bd9b7454458d Mon Sep 17 00:00:00 2001
From: Hannes Frederic Sowa <hannes@redhat.com>
Date: Mon, 8 Dec 2014 17:33:03 -0500
Subject: [net] udp: do not report ICMP redirects to user space

Message-id: <5a2e0e856700d19a2d7a17090fc744072f88ea3d.1418058212.git.hannes@redhat.com>
Patchwork-id: 101181
O-Subject: [RHEL7.1 net 07/13] net: udp: do not report ICMP redirects to user space
Bugzilla: 1170116
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Flavio Leitner <fbl@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1170116
Upstream Status: 1a462d189280b560bd84af1407e4d848e262b3b3
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8344420

Upstream Commit:
commit 1a462d189280b560bd84af1407e4d848e262b3b3
Author: Duan Jiong <duanj.fnst@cn.fujitsu.com>
Date:   Fri Sep 20 18:20:28 2013 +0800

    net: udp: do not report ICMP redirects to user space

    Redirect isn't an error condition, it should leave
    the error handler without touching the socket.

    Signed-off-by: Duan Jiong <duanj.fnst@cn.fujitsu.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Hannes Frederic Sowa <hannes@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/udp.c b/net/ipv4/udp.c
index e75dd5d..a66d0e6 100644
--- a/net/ipv4/udp.c
+++ b/net/ipv4/udp.c
@@ -659,7 +659,7 @@ void __udp4_lib_err(struct sk_buff *skb, u32 info, struct udp_table *udptable)
   break;
  case ICMP_REDIRECT:
   ipv4_sk_redirect(skb, sk);
-  break;
+  goto out;
  }
 
  /*
diff --git a/net/ipv6/udp.c b/net/ipv6/udp.c
index 4e89e0f..1db307c 100644
--- a/net/ipv6/udp.c
+++ b/net/ipv6/udp.c
@@ -517,8 +517,10 @@ void __udp6_lib_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
 
  if (type == ICMPV6_PKT_TOOBIG)
   ip6_sk_update_pmtu(skb, sk, info);
- if (type == NDISC_REDIRECT)
+ if (type == NDISC_REDIRECT) {
   ip6_sk_redirect(skb, sk);
+  goto out;
+ }
 
  np = inet6_sk(sk);
 
-- 
1.7.1